File: TLX/techfest-2024/B.py
def binexp(a, b, MOD):
    res = 1
    while b > 0:
        if b & 1:
            res = res * a % MOD
        a = a * a % MOD
        b >>= 1
    return res

def main():
    a, b, c, n = map(int, input().split())
    if n == 1:
        print(0)
        return
    bc_mod = binexp(b, c, n-1)
    result = binexp(a, bc_mod, n)
    print(result)

if __name__ == "__main__":
    main()

File: TLX/techfest-2024/B.cpp
#include <bits/stdc++.h>

using namespace std;

#define int unsigned long long

int binexp(int a, int b, int MOD){
    int res = 1;
    while(b > 0){
        if(b & 1) res = res * a % MOD;
        a = a * a % MOD;
        b >>= 1;
    }
    return res;
}

signed main(){
    ios::sync_with_stdio(0);
    cin.tie(0);
    int a,b,c,n; cin >> a >> b >> c >> n;
    if (n == 1){
        cout << 0;
        return 0;
    }
    cout << binexp(a, binexp(b, c, n-1), n);
}


/*
a^b^c mod n = pow(a, pow(b, c, n-1), n)
*/